[runtimes] Fix parsing of LIB{CXX,CXXABI,UNWIND}_TEST_PARAMS

arichardson opened the pull request:
Since 78d649a417b48cb8a2ba2e755f0e7c8fb8b1bb83 the recommended way to pass an executor is to use the _TEST_PARAMS variable, which means we now pass more complicated value (including ones that may contain multiple `=`) as part of this variable. However, the `REGEX REPLACE` being used has greedy matches so everything up to the last = becomes part of the variable name which results in invalid syntax in the generated lit config file.

This was noticed due to builder failures for those using the CrossWinToARMLinux.cmake cache file.

arichardson commented:
@vvereschaka I tested this locally by adding a parameter with quotes to my CMake invocation, so it should hopefully work. Could you please test if this fixes the builder?

ldionne commented on runtimes/cmake/Modules/HandleLitArguments.cmake:
```diff
@@ -0,0 +1,20 @@
+
+macro(serialize_lit_param param value)
+  string(APPEND SERIALIZED_LIT_PARAMS "config.${param} = ${value}\n")
```
This looks reasonable, but could we avoid using the global variable `SERIALIZED_LIT_PARAMS` here? We should thread it through the functions instead, otherwise these functions all "return their result" by modifying the global variable, which is yucky.

vvereschaka commented on clang/cmake/caches/CrossWinToARMLinux.cmake:
```diff
@@ -151,6 +151,10 @@ set(RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_LIBCXX_ENABLE_NEW_DELETE_DEFINITIONS
+set(RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_LIBUNWIND_TEST_PARAMS "${RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_TEST_PARAMS}" CACHE INTERNAL "")
+set(RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_LIBCXXABI_TEST_PARAMS "${RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_TEST_PARAMS}" CACHE INTERNAL "")
+set(RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_LIBCXX_TEST_PARAMS "${RUNTIMES_${TOOLCHAIN_TARGET_TRIPLE}_TEST_PARAMS}" CACHE INTERNAL "")
+
```
It is not working. This will store the empty variables in the cache file. If you check your local `CMakeCache.txt` for these variables you'll find them empty. You need to cache them with the values at the end of preparation (please check my original diff file).

arichardson replied:
Should be fixed in the latest version.

arichardson committed:
[runtimes] Factor out serialize_lit_param to a separate file

This introduces new macros that expect an output variable and a new
macro serialize_lit_string_param() that ensure that the serialized
value is escaped appropriately to store it into a python change.

arichardson committed:
[runtimes] Fix parsing of LIB{CXX,CXXABI,UNWIND}_TEST_PARAMS

Since 78d649a417b48cb8a2ba2e755f0e7c8fb8b1bb83 the recommended way to pass
an executor is to use the _TEST_PARAMS variable, which means we now pass
more complicated value (including ones that may contain multiple `=`) as
part of this variable. However, the `REGEX REPLACE` being used has greedy
matches so everything up to the last = becomes part of the variable name
which results in invalid syntax in the generated lit config file.

This was noticed due to builder failures for those using the
CrossWinToARMLinux.cmake cache file.

Co-authored-by: Vladimir Vereschaka <vvereschaka@accesssoftek.com>

vvereschaka commented:
@arichardson,
any progress?!

arichardson commented:
> @arichardson, any progress?!

Sorry thought I'd pushed the latest version earlier, turns out there was an error during the push. Updated now.

vvereschaka commented:
@arichardson,
thank you. I'm going to test it during the next 15-20 minutes.

vvereschaka commented:
@arichardson ,
works fine on the builders locally. Thank you.
Would yo merge these changes if they are ready?

arichardson commented:
> @arichardson , works fine on the builders locally. Thank you. Would yo merge these changes if they are ready?

Thanks for testing! I'll merge these changes as soon as @ldionne is happy with the new macros. Once that happens I'll push the first commit manually and then merge the second one via the github UI (since we can't do "rebase and merge" due to repo settings).

vvereschaka commented:
@ldionne @arichardson any progress?

vvereschaka commented:
@ldionne @arichardson ,
the problem is still there for more than 10 days already. Do you guys plan to fix it!

arichardson commented:
> @ldionne @arichardson , the problem is still there for more than 10 days already. Do you guys plan to fix it!

I am waiting for @ldionne to approve this PR.

vvereschaka commented:
@arichardson ,
would you commit these changes without @ldionne 's approval? or revert your #66545?

ldionne commented:
Thanks for the fix and sorry for the delay!
